File: e2e-tests/cmd/go/procfs.go
// SPDX-License-Identifier: MPL-2.0
/*
 * libpathrs: safe path resolution on Linux
 * Copyright (C) 2019-2025 SUSE LLC
 * Copyright (C) 2026 Aleksa Sarai <cyphar@cyphar.com>
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package main

import (
	"context"
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/urfave/cli/v3"
	"golang.org/x/sys/unix"

	"cyphar.com/go-pathrs/procfs"
)

type procfsOpenFunc func(path string, flags int) (*os.File, error)

var procfsCmd = &cli.Command{
	Name:  "procfs",
	Usage: "ProcfsHandle::* operations",
	Flags: []cli.Flag{
		&cli.BoolFlag{
			Name:  "unmasked",
			Usage: "use unmasked procfs handle",
		},
		&cli.StringFlag{
			Name:  "base",
			Usage: "base path for procfs operations (root, pid=<n>, self, thread-self)",
			Value: "root",
		},
	},
	Commands: []*cli.Command{
		procfsOpenCmd,
		procfsReadlinkCmd,
	},

	Before: func(ctx context.Context, cmd *cli.Command) (_ context.Context, Err error) {
		var opts []procfs.OpenOption
		if cmd.Bool("unmasked") {
			opts = append(opts, procfs.UnmaskedProcRoot)
		}
		proc, err := procfs.Open(opts...)
		if err != nil {
			return nil, err
		}
		ctx = context.WithValue(ctx, "procfs", proc)
		defer func() {
			if Err != nil {
				_ = proc.Close()
			}
		}()

		return ctx, nil
	},

	After: func(ctx context.Context, cmd *cli.Command) error {
		if proc, ok := ctx.Value("procfs").(*procfs.Handle); ok {
			_ = proc.Close()
		}
		return nil
	},
}

var procfsOpenCmd = cmdWithOptions(&cli.Command{
	Name:  "open",
	Usage: "open a path in procfs",
	Flags: []cli.Flag{
		&cli.BoolWithInverseFlag{
			Name:  "follow",
			Usage: "follow trailing symlinks",
			Value: true,
		},
	},
	Arguments: []cli.Argument{
		&cli.StringArg{
			Name: "subpath",
		},
	},

	Action: func(ctx context.Context, cmd *cli.Command) error {
		proc := ctx.Value("procfs").(*procfs.Handle)
		follow := cmd.Bool("follow")
		subpath := cmd.StringArg("subpath")

		oflags := unix.O_RDONLY
		if val := ctx.Value("oflags"); val != nil {
			oflags = val.(int)
		}
		if !follow {
			oflags |= unix.O_NOFOLLOW
		}

		var (
			f   *os.File
			err error
		)

		baseStr := cmd.String("base")
		if pidStr, ok := strings.CutPrefix(baseStr, "pid="); ok {
			var pid int
			pid, err = strconv.Atoi(pidStr)
			if err != nil {
				return fmt.Errorf("failed to parse --base=%q: %w", pidStr, err)
			}
			f, err = proc.OpenPid(pid, subpath, oflags)
		} else {
			switch baseStr {
			case "root":
				f, err = proc.OpenRoot(subpath, oflags)
			case "self":
				f, err = proc.OpenSelf(subpath, oflags)
			case "thread-self":
				var closer procfs.ThreadCloser
				f, closer, err = proc.OpenThreadSelf(subpath, oflags)
				if closer != nil {
					defer closer()
				}
			default:
				return fmt.Errorf("invalid --base value %q", baseStr)
			}
		}
		if err != nil {
			return err
		}
		defer f.Close()

		fmt.Println("FILE-PATH", f.Name())
		// TODO: Input/output file data.
		return nil
	},
}, oflags("oflags", "O_* flags to use when opening the file", unix.O_RDONLY))

var procfsReadlinkCmd = &cli.Command{
	Name:  "readlink",
	Usage: "read the target path of a symbolic link in procfs",
	Arguments: []cli.Argument{
		&cli.StringArg{
			Name: "subpath",
		},
	},

	Action: func(ctx context.Context, cmd *cli.Command) error {
		proc := ctx.Value("procfs").(*procfs.Handle)
		subpath := cmd.StringArg("subpath")

		var base procfs.ProcBase
		baseStr := cmd.String("base")
		if pidStr, ok := strings.CutPrefix(baseStr, "pid="); ok {
			pid, err := strconv.Atoi(pidStr)
			if err != nil {
				return fmt.Errorf("failed to parse --base=%q: %w", pidStr, err)
			}
			base = procfs.ProcPid(pid)
		} else {
			switch baseStr {
			case "root":
				base = procfs.ProcRoot
			case "self":
				base = procfs.ProcSelf
			case "thread-self":
				base = procfs.ProcThreadSelf
			default:
				return fmt.Errorf("invalid --base value %q", baseStr)
			}
		}

		target, err := proc.Readlink(base, subpath)
		if err != nil {
			return err
		}
		fmt.Println("LINK-TARGET", target)
		return nil
	},
}

File: e2e-tests/cmd/rust/src/procfs.rs
// SPDX-License-Identifier: MPL-2.0
/*
 * libpathrs: safe path resolution on Linux
 * Copyright (C) 2019-2025 SUSE LLC
 * Copyright (C) 2026 Aleksa Sarai <cyphar@cyphar.com>
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

use crate::utils;
use pathrs::{
    flags::OpenFlags,
    procfs::{ProcfsBase, ProcfsHandleBuilder, ProcfsHandleRef},
};

use std::{ffi::OsStr, path::PathBuf};

use anyhow::{anyhow, Error};
use clap::{
    builder::TypedValueParser, error::ErrorKind as ClapErrorKind, Arg, ArgAction, ArgMatches,
    Command,
};

#[derive(Debug, Clone, Copy)]
struct ProcfsBaseParser;

impl TypedValueParser for ProcfsBaseParser {
    type Value = ProcfsBase;

    fn parse_ref(
        &self,
        _cmd: &Command,
        _arg: Option<&Arg>,
        value: &OsStr,
    ) -> Result<Self::Value, clap::Error> {
        let value = value.to_str().ok_or_else(|| {
            clap::Error::raw(
                ClapErrorKind::InvalidUtf8,
                "procfs base contained invalid utf8 characters",
            )
        })?;

        match (value, value.strip_prefix("pid=")) {
            (_, Some(pid)) => Ok(ProcfsBase::ProcPid(pid.parse().map_err(|err| {
                clap::Error::raw(
                    ClapErrorKind::ValueValidation,
                    format!("{value} is an invalid octal mode: {err:?}"),
                )
            })?)),
            ("root", _) => Ok(ProcfsBase::ProcRoot),
            ("self", _) => Ok(ProcfsBase::ProcSelf),
            ("thread-self", _) => Ok(ProcfsBase::ProcThreadSelf),
            (value, None) => Err(clap::Error::raw(
                ClapErrorKind::ValueValidation,
                format!("{value} is an invalid procfs base"),
            )),
        }
    }
}

fn open_cli() -> Command {
    Command::new("open")
        .about("open a path in procfs")
        .arg(
            utils::oflags_arg("oflags", "O_* flags to use when opening the file")
                .long("oflags")
                .default_value("O_RDONLY"),
        )
        .args(utils::toggle_arg("follow", "follow trailing symlinks"))
        .arg(
            utils::subpath_arg("subpath")
                .help("path inside procfs base")
                .required(true),
        )
}

fn open(procfs: ProcfsHandleRef<'_>, base: ProcfsBase, matches: &ArgMatches) -> Result<(), Error> {
    let subpath = matches
        .get_one::<PathBuf>("subpath")
        .expect("subpath should always be set");
    let follow = *matches
        .get_one::<bool>("follow")
        .expect("follow should be set");
    let oflags = *matches
        .get_one::<OpenFlags>("oflags")
        .expect("oflags should always be set");

    let f = if follow {
        procfs.open_follow(base, subpath, oflags)
    } else {
        procfs.open(base, subpath, oflags)
    }?;
    utils::print_file(&f)?;

    Ok(())
}

fn readlink_cli() -> Command {
    Command::new("readlink")
        .about("read the target path of a symbolic link in procfs")
        .arg(
            utils::subpath_arg("subpath")
                .help("path inside procfs base")
                .required(true),
        )
}

fn readlink(
    procfs: ProcfsHandleRef<'_>,
    base: ProcfsBase,
    matches: &ArgMatches,
) -> Result<(), Error> {
    let subpath = matches
        .get_one::<PathBuf>("subpath")
        .expect("subpath should always be set");

    let link_target = procfs.readlink(base, subpath)?;
    println!("LINK-TARGET {}", link_target.to_string_lossy());
    Ok(())
}

pub(crate) fn cli() -> Command {
    Command::new("procfs")
        .about("ProcfsHandle::* operations")
        .arg(
            Arg::new("unmasked")
                .long("unmasked")
                .help("use unmasked procfs handle")
                .action(ArgAction::SetTrue),
        )
        .arg(
            Arg::new("base")
                .long("base")
                .help("base path for procfs operations (root, pid=<n>, self, thread-self)")
                .value_name("PROC_*")
                .default_value("root")
                .value_parser(ProcfsBaseParser)
                .action(ArgAction::Set),
        )
        .subcommand(open_cli())
        .subcommand(readlink_cli())
}

pub(crate) fn subcommand(matches: &ArgMatches) -> Result<(), Error> {
    let procfs = {
        let mut b = ProcfsHandleBuilder::new();

        // MSRV(1.85): Use let chain here (Rust 2024).
        if matches.get_one::<bool>("unmasked") == Some(&true) {
            b.set_unmasked();
        }

        b.build()
    }?;
    let base = *matches
        .get_one::<ProcfsBase>("base")
        .expect("base should always be set");

    match matches.subcommand() {
        Some(("open", sub_matches)) => open(procfs, base, sub_matches),
        Some(("readlink", sub_matches)) => readlink(procfs, base, sub_matches),
        Some((subcommand, _)) => {
            // We should never end up here.
            Err(anyhow!("unknown 'procfs' subcommand '{subcommand}'"))
        }
        None => Err(anyhow!("no 'procfs' subcommand specified")),
    }
}
